### ISSUE TITLE
Are rewrites confluent? Tester currently assumes confluence, this may be wrong.

### ISSUE THREAD

**Issue by niklasdewally** [conjure-cp/conjure-oxide#223]
We currently do a check against a "final rewritten model" in the integration tester. However, if the system is not confluent (i.e. we can get different simplest forms for a given model), this may break the tester which expects a single form of "rewritten" model.

As we are just applying rules all the way down without any set process (compared to savile row etc), I don't think confluence is realistic here, so we should not assume it in testing.

TL;DR
1. Do rewrites of a model always give the same output?
2. Can we fix the integration testds for rewriting to not assume this? 

@ozgurakgun @gskorokhod

**Comment by gskorokhod** [comment]
I agree. I just implemented it as a quick way to test the rewriter, but you're right.
An alternative approach would be to only check the original parsed model and the solutions, and for the rewrite step, instead of comparing output to the "one single correct rewritten model", we can walk the tree and do some sanity checks (I.e: there is no Nothings, no unexpected constraints, no applicable rules left, etc)

Do we have a type checking step for the AST btw? If we do (or if it's in the works), we can just do that

**Comment by niklasdewally** [comment]
> An alternative approach would be to only check the original parsed model and the solutions, and for the rewrite step, instead of comparing output to the "one single correct rewritten model", we can walk the tree and do some sanity checks (I.e: there is no Nothings, no unexpected constraints, no applicable rules left, etc)

I'll wait for Oz for an answer on confluence, but I agree that if we don't have it this would be a good way to test 👍🏻 

Your solution is also presuming that each transformation of the model is valid essence (excepting bubbles, which are explicitly a type error) - I discussed this with felix recently, but I don't think this property has been decided either yet? It seems sensible, but it's not a given.

**Comment by niklasdewally** [comment]
We will need type and well-formedness checks anyways I imagine.

**Comment by ozgurakgun** [comment]
We want them to be confluent (given we make the same decisions when multiple rules are applicable).

For now only 1 rule is applicable (if we are lucky) so absolutely should be confluent.

**Comment by ozgurakgun** [comment]
Bubbles don't have to be a type error!

`{ x @ C }` has the same type as `x`, provided `C` is a well-typed constraint (i.e. bool).

**Comment by niklasdewally** [comment]
> For now only 1 rule is applicable (if we are lucky) so absolutely should be confluent.

By this do you mean that we should try to be confluent forever, or that we should just revisit this later when we actually haven enough rules to matter? May impact how we design priority and rulesets and all that...

**Comment by ozgurakgun** [comment]
It should be confluent forever.

Importantly: confluent **given the same answers to all multiple choice questions**. Which is probably a stretch of the definition of what confluence means. I can explain this further tomorrow if it's not clear.

**Comment by niklasdewally** [comment]
> It should be confluent forever.
> 
> Importantly: confluent **given the same answers to all multiple choice questions**. Which is probably a stretch of the definition of what confluence means. I can explain this further tomorrow if it's not clear.

That makes sense to me :) 

One immediate problem is we don't have a guaranteed rule ordering when using linkme/inventory/stuff like that - if we want determinism and confluence we should probably give them a lexical ordering (or similar) so that this doesn't randomly break in weird and subtle ways


@lixitrixi
